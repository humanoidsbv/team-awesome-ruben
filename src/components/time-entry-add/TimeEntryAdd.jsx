import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';

import IconPlus from '../../assets/icons/icon-plus.svg';
import styles from './TimeEntryAdd.module.css';

const TimeEntryAdd = ({ addFormData }) => {
  const today = new Date()
    .toISOString()
    .split('T')
    .shift();

  const [client, setClient] = useState('');
  const [activity, setActivity] = useState('');
  const [date, setDate] = useState(today);
  const [startTime, setStartTime] = useState('09:00');
  const [endTime, setEndTime] = useState('17:30');

  const [validity, setValidity] = useState({});
  const [isFormVisible, setIsFormVisible] = useState(true);

  const formRef = useRef(null);

  const handleToggle = () => setIsFormVisible(!isFormVisible);

  const handleBlur = event => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const handleSubmit = event => {
    event.preventDefault();

    addFormData({
      client,
      id: Math.random(),
      startTimestamp: new Date(`${date} ${startTime}`).toISOString(),
      stopTimestamp: new Date(`${date} ${endTime}`).toISOString()
    });

    setClient('');
    setActivity('');
  };

  return (
    <div className={`${styles.container}`}>
      <form
        className={` ${styles.form} ${isFormVisible && styles.formVisible}`}
        onSubmit={handleSubmit}
        ref={formRef}
      >
        <button
          className={styles.closeButton}
          onClick={handleToggle}
          type="button"
        >
          <IconPlus />
        </button>

        <label
          className={`${styles.label} ${styles.labelLarge}`}
          htmlFor="client"
        >
          <p className={styles.labelText}>CLIENT</p>
          <input
            className={`${styles.input} ${
              validity.client === false ? styles.invalidInput : ''
            }`}
            id="client"
            maxLength="30"
            minLength="2"
            name="client"
            onBlur={handleBlur}
            onChange={({ target }) => setClient(target.value)}
            value={client}
            required
          />
        </label>

        <label
          className={`${styles.label} ${styles.labelLarge}`}
          htmlFor="activity"
        >
          <p className={styles.labelText}>ACTIVITY</p>
          <input
            className={`${styles.input} ${
              validity.activity === false ? styles.invalidInput : ''
            }`}
            id="activity"
            maxLength="30"
            minLength="2"
            name="activity"
            onBlur={handleBlur}
            onChange={({ target }) => setActivity(target.value)}
            required
            value={activity}
          />
        </label>
        <label
          className={`${styles.label} ${styles.labelMedium} ${styles.date}`}
          htmlFor="date"
        >
          <p className={styles.labelText}>DATE</p>
          <input
            className={styles.input}
            id="date"
            name="date"
            onChange={({ target }) => setDate(target.value)}
            required
            type="date"
            value={date}
          />
        </label>

        <label
          className={`${styles.label} ${styles.labelSmall} ${styles.labelHalfWidth}`}
          htmlFor="startTime"
        >
          <p className={styles.labelText}>FROM</p>
          <input
            className={`${styles.input} ${styles.timeStamp}`}
            id="startTime"
            name="startTime"
            onChange={({ target }) => setStartTime(target.value)}
            type="time"
            value={startTime}
          />
        </label>
        <label
          className={`${styles.label} ${styles.labelSmall} ${styles.labelHalfWidth}`}
          htmlFor="endTime"
        >
          <p className={styles.labelText}>TO</p>
          <input
            className={`${styles.input} ${styles.timeStamp}`}
            id="endTime"
            name="endTime"
            onChange={({ target }) => setEndTime(target.value)}
            type="time"
            value={endTime}
          />
        </label>
        <button
          className={styles.addButton}
          disabled={validity.form !== true}
          type="submit"
        >
          Add
        </button>
      </form>

      {!isFormVisible && (
        <button
          className={styles.showForm}
          type="button"
          onClick={handleToggle}
        >
          <IconPlus className={styles.showFormIcon} />
          New time entry
        </button>
      )}
    </div>
  );
};

TimeEntryAdd.propTypes = {
  addFormData: PropTypes.func.isRequired
};

export default TimeEntryAdd;
