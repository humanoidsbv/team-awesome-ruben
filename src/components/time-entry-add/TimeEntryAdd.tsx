import React, { useState, useRef } from 'react';

import IconPlus from '../../assets/icons/icon-plus.svg';
import styles from './TimeEntryAdd.module.css';

import { ValidityState } from '../../../shared/types';

const TimeEntryAdd = ({ addFormData, clients }): React.ReactElement => {
  const today = new Date()
    .toISOString()
    .split('T')
    .shift();

  const [client, setClient] = useState('');
  const [activity, setActivity] = useState('');
  const [date, setDate] = useState(today);
  const [startTime, setStartTime] = useState('09:00');
  const [endTime, setEndTime] = useState('17:30');

  const [validity, setValidity] = useState<ValidityState>({});
  const [isFormVisible, setIsFormVisible] = useState(true);

  const formRef = useRef(null);

  const handleToggle = (): void => setIsFormVisible(!isFormVisible);

  const handleBlur = (
    event: React.ChangeEvent<HTMLInputElement | HTMLSelectElement>
  ): void => {
    setValidity({
      ...validity,
      form: formRef.current.checkValidity(),
      [event.target.name]: event.target.checkValidity()
    });
  };

  const handleSubmit = (event): void => {
    event.preventDefault();

    addFormData({
      client: Number(client),
      id: Math.random(),
      startTimestamp: new Date(`${date} ${startTime}`).toISOString(),
      stopTimestamp: new Date(`${date} ${endTime}`).toISOString()
    });

    setClient('');
    setActivity('');
  };

  return (
    <div className={`${styles.container}`}>
      <form
        className={`${styles.form} ${isFormVisible && styles.formVisible}`}
        onSubmit={handleSubmit}
        ref={formRef}
      >
        <button
          className={styles.closeButton}
          onClick={handleToggle}
          type="button"
        >
          <IconPlus />
        </button>

        <label
          className={`${styles.label} ${styles.labelLarge}`}
          htmlFor="client"
        >
          <p className={styles.labelText}>CLIENT</p>
          <select
            className={`${styles.input} ${
              validity.client === false ? styles.invalidInput : ''
            }`}
            name="client"
            onBlur={handleBlur}
            onChange={({
              target
            }: React.ChangeEvent<HTMLSelectElement>): void =>
              setClient(target.value)
            }
            value={client}
            required
          >
            <option value="" disabled>
              -- Select a client --
            </option>
            {clients.map(
              ({ companyName, id }: { companyName: string; id: number }) => (
                <option value={id} key={id}>
                  {companyName}
                </option>
              )
            )}
          </select>
        </label>

        <label
          className={`${styles.label} ${styles.labelLarge}`}
          htmlFor="activity"
        >
          <p className={styles.labelText}>ACTIVITY</p>
          <input
            className={`${styles.input} ${
              validity.activity === false ? styles.invalidInput : ''
            }`}
            maxLength={30}
            minLength={2}
            name="activity"
            onBlur={handleBlur}
            onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void =>
              setActivity(target.value)
            }
            required
            value={activity}
          />
        </label>
        <label
          className={`${styles.label} ${styles.labelMedium} ${styles.date}`}
          htmlFor="date"
        >
          <p className={styles.labelText}>DATE</p>
          <input
            className={styles.input}
            name="date"
            onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void =>
              setDate(target.value)
            }
            required
            type="date"
            value={date}
          />
        </label>

        <label
          className={`${styles.label} ${styles.labelSmall} ${styles.labelHalfWidth}`}
          htmlFor="startTime"
        >
          <p className={styles.labelText}>FROM</p>
          <input
            className={`${styles.input} ${styles.timeStamp}`}
            name="startTime"
            onChange={({ target }: React.ChangeEvent<HTMLInputElement>): void =>
              setStartTime(target.value)
            }
            type="time"
            value={startTime}
          />
        </label>
        <label
          className={`${styles.label} ${styles.labelSmall} ${styles.labelHalfWidth}`}
          htmlFor="endTime"
        >
          <p className={styles.labelText}>TO</p>
          <input
            className={`${styles.input} ${styles.timeStamp}`}
            name="endTime"
            onChange={({
              target
            }: React.ChangeEvent<HTMLInputElement>): void | string =>
              setEndTime(target.value)
            }
            type="time"
            value={endTime}
          />
        </label>
        <button
          className={styles.addButton}
          disabled={validity.form !== true}
          type="submit"
        >
          Add
        </button>
      </form>

      {!isFormVisible && (
        <button
          className={styles.showForm}
          type="button"
          onClick={handleToggle}
        >
          <IconPlus className={styles.showFormIcon} />
          New time entry
        </button>
      )}
    </div>
  );
};

export default TimeEntryAdd;
