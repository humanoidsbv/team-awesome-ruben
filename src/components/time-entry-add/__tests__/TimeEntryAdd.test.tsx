import * as React from 'react';
import { shallow } from 'enzyme';

import TimeEntryAdd from '../TimeEntryAdd';

const ClientMockedData = {
  branch: '123',
  branchAddress: '123',
  companyName: '123',
  id: 123,
  locality: '123',
  postalCode: '123'
};

it('matches output with  snapshot', () => {
  const timeEntryAdd = shallow(
    <TimeEntryAdd addFormData={({}) => {}} clients={[ClientMockedData]} />
  );
  expect(timeEntryAdd).toMatchSnapshot();
});
