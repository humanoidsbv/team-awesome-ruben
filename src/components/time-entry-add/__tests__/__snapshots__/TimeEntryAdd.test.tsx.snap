// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches output with  snapshot 1`] = `
<div
  className="container"
>
  <form
    className="form formVisible"
    onSubmit={[Function]}
  >
    <button
      className="closeButton"
      onClick={[Function]}
      type="button"
    >
      <test-file-stub />
    </button>
    <label
      className="label labelLarge"
      htmlFor="client"
    >
      <p
        className="labelText"
      >
        CLIENT
      </p>
      <select
        className="input "
        name="client"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        value=""
      >
        <option
          disabled={true}
          value=""
        >
          -- Select a client --
        </option>
        <option
          key="123"
          value={123}
        >
          123
        </option>
      </select>
    </label>
    <label
      className="label labelLarge"
      htmlFor="activity"
    >
      <p
        className="labelText"
      >
        ACTIVITY
      </p>
      <input
        className="input "
        maxLength={30}
        minLength={2}
        name="activity"
        onBlur={[Function]}
        onChange={[Function]}
        required={true}
        value=""
      />
    </label>
    <label
      className="label labelMedium date"
      htmlFor="date"
    >
      <p
        className="labelText"
      >
        DATE
      </p>
      <input
        className="input"
        name="date"
        onChange={[Function]}
        required={true}
        type="date"
        value="2019-10-21"
      />
    </label>
    <label
      className="label labelSmall labelHalfWidth"
      htmlFor="startTime"
    >
      <p
        className="labelText"
      >
        FROM
      </p>
      <input
        className="input timeStamp"
        name="startTime"
        onChange={[Function]}
        type="time"
        value="09:00"
      />
    </label>
    <label
      className="label labelSmall labelHalfWidth"
      htmlFor="endTime"
    >
      <p
        className="labelText"
      >
        TO
      </p>
      <input
        className="input timeStamp"
        name="endTime"
        onChange={[Function]}
        type="time"
        value="17:30"
      />
    </label>
    <button
      className="addButton"
      disabled={true}
      type="submit"
    >
      Add
    </button>
  </form>
</div>
`;
